ads/google/a4a/line-delimited-response-handler.js

```diff
@@ -0,0 +1,164 @@
+             pos = hasDelimiter ? delimIndex + 1 : end;
+             if (hasDelimiter) {
+               if (!metaData) {
+                 metaData = safeJsonParse_(snippet);
+               } else {
+                 // Have metaData/HTML so we can call chunkHandler and reset.
+                 chunkHandler(
+                   unescapeHtml_(snippet),
+                   /** @type {!Object<string, *>} */(metaData),
+                   done && end == chunk.length);
+                 metaData = undefined;
+               }
+               snippet = '';
+             } else if (countTrailingSlashes(chunk, pos - 1) & 1) {
+               // Decrement 'end' to skip the last character of the chunk if it
+               // ends with an odd number of slashes, since that implies that
+               // the last slash is used to escape a character that we haven't
+               // received yet. So we skip the last byte in this iteration and
+               // wait for the next chunk to see what is being escaped.
+               end--;
+             }
+             pos = hasDelimiter ? delimIndex + 1 : end;
+           } while (hasDelimiter && pos < chunk.length);
+         };
+         handleFetchResponseStream_(
+           response.body.getReader(), chunkCallback, new TextDecoder('utf-8'));
+       }
+       return response;
+     });
+ }
+
+ /**
+  * @param {!ReadableStreamReader} reader The reader of the fetch's
+  *  response stream.
+  * @param {!function(string, boolean)} chunkHandler
+  * @param {!TextDecoder} textDecoder
+  * @private
+  */
+ function handleFetchResponseStream_(reader, chunkHandler, textDecoder) {
+   return reader.read().then(result => {
+     if (result.value) {
+       chunkHandler(
+         textDecoder.decode(
+           /** @type {!ArrayBuffer} */(result.value), {'stream': true}),
+         result.done);
+     }
+     if (!result.done) {
+       // More chunks to read.
+       handleFetchResponseStream_.apply(this, arguments);
+     }
+   });
+ }
+
+ /**
+  * @param {string} text
+  * @param {!function(string, !Object<string, *>, boolean)} chunkHandler
+  * @visibleForTesting
+  */
+ function chunkHandleFullResponse(text, chunkHandler) {
+   const lines = text.split('\n');
+   // Note that its expected for an extra return to existing in the format.
+   let linesRemaining = lines.length;
+   let metaData;
+   lines.forEach(line => {
+     if (!metaData) {
+       metaData = safeJsonParse_(line);
+     } else {
+       chunkHandler(unescapeHtml_(line),
+          /** @type {!Object<string,*>} */(metaData), !(--linesRemaining));
+       metaData = undefined;
+     }
+   });
+ }
+
+ /**
+  * Unescapes characters that are escaped in line-delimited JSON-HTML.
+  * @param {string} html An html snippet.
+  * @return {string}
+  * @private
+  */
+ function unescapeHtml_(html) {
+   return html.replace(
+     /\\(n|r|\\)/g,
+     (_, match) => match == 'n' ? '\n' : match == 'r' ? '\r' : '\\');
```

Comment:
Its possible the creative is some magical string plus javascript expecting the magic string to have \r characters so better to ensure it exactly matches what was send over the wire.